File: backend/schemas/traceability.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

TraceabilityStatus = Literal["covered", "failed", "skipped", "not_run"]


class TraceabilityTestCaseItem(BaseModel):
    id: int
    code: str
    title: str
    status: TraceabilityStatus
    run_id: int | None = None
    executed_at: datetime | None = None


class TraceabilityRequirementItem(BaseModel):
    id: int
    req_id: str
    version: int
    title: str
    module: str | None
    status: str
    is_uncovered: bool
    coverage_percent: float
    test_cases: list[TraceabilityTestCaseItem]


class TraceabilityResponse(BaseModel):
    items: list[TraceabilityRequirementItem]

File: frontend/src/lib/traceability.ts
import { authFetch } from './api'

export type TraceabilityStatus = 'covered' | 'failed' | 'skipped' | 'not_run'

export type TraceabilityTestCaseItem = {
  id: number
  code: string
  title: string
  status: TraceabilityStatus
  execution_id: number | null
  executed_at: string | null
  release_id: number | null
  release_version_name: string | null
}

export type TraceabilityRequirementItem = {
  id: number
  req_id: string
  version: number
  title: string
  module: string | null
  status: string
  is_uncovered: boolean
  coverage_percent: number
  test_cases: TraceabilityTestCaseItem[]
}

export type TraceabilityResponse = {
  items: TraceabilityRequirementItem[]
}

export async function getTraceability(projectId: number): Promise<TraceabilityResponse> {
  return authFetch<TraceabilityResponse>(`/traceability?project_id=${projectId}`)
}
